dockstore-webservice/src/main/java/io/dockstore/webservice/resources/EntryResource.java

```diff
@@ -145,11 +150,43 @@ public Entry addAliases(@ApiParam(hidden = true) @Parameter(hidden = true, name
     public List<VersionVerifiedPlatform> getVerifiedPlatforms(@Parameter(hidden = true, name = "user")@Auth Optional<User> user,
             @Parameter(name = "entryId", description = "id of the entry", required = true, in = ParameterIn.PATH) @PathParam("entryId") Long entryId) {
         Entry<? extends Entry, ? extends Version> entry = toolDAO.getGenericEntryById(entryId);
-        checkOptionalAuthRead(user, entry);
+        checkEntry(entry);
+
+        if (!entry.getIsPublished()) {
+            checkUser(user.get(), entry);
+        }
+
         List<VersionVerifiedPlatform> verifiedVersions = versionDAO.findEntryVersionsWithVerifiedPlatforms(entryId);
         return verifiedVersions;
     }
 
+    @GET
+    @Timed
+    @UnitOfWork(readOnly = true)
+    @Path("/{entryId}/versions/{versionId}/fileTypes")
+    @ApiOperation(value = "Retrieve the file types of a version's sourcefiles",  hidden = true)
+    @Operation(operationId = "getVersionsFileTypes", description = "Retrieve the file types of a version's sourcefile", security = @SecurityRequirement(name = OPENAPI_JWT_SECURITY_DEFINITION_NAME))
+    public SortedSet<DescriptorLanguage.FileType> getVersionsFileTypes(@Parameter(hidden = true, name = "user")@Auth Optional<User> user,
+            @Parameter(name = "entryId", description = "Entry to retrieve the version from", required = true, in = ParameterIn.PATH) @PathParam("entryId") Long entryId,
+            @Parameter(name = "versionId", description = "Version to retrieve the sourcefile types from", required = true, in = ParameterIn.PATH) @PathParam("versionId") Long versionId) {
+        Entry<? extends Entry, ? extends Version> entry = toolDAO.getGenericEntryById(entryId);
+        checkEntry(entry);
+
+        if (!entry.getIsPublished()) {
+            checkUser(user.get(), entry);
+        }
+
+        Version version = versionDAO.findVersionInEntry(entryId, versionId);
+        if (version == null) {
+            throw new CustomWebApplicationException("Version " + versionId + " does not exist for this entry", HttpStatus.SC_BAD_REQUEST);
+        }
+
+        SortedSet<DescriptorLanguage.FileType> fileTypes = new TreeSet<>();
+        SortedSet<SourceFile> sourceFiles = version.getSourceFiles();
+        fileTypes = sourceFiles.stream().map(sourceFile -> sourceFile.getType()).collect(Collectors.toCollection(TreeSet::new));
```

Comment:
1. This is unfortunate, but may be fast enough for a single version, so probably not worth optimizing yet.
2. Why do you use a `TreeSet` instead of `Collectors.toSet`? TreeSet sorts, but for an enum it's just going to sort on the order of declaration, I think, and I don't think that order is significant. I could be wrong though.

Reply:
On further reflection, it probably is better to use TreeSet. Although the particular  ordering doesn’t really matter much, the fact that it will return the same order every time is a good thing.